Feature/java long to string

owen-jones-diffblue opened the pull request:
Didn't manage to make binary fast enough to be turned on to core. There are two competing approaches: constructive (construct the number as a formula involving the string) and non-constructive (have a separate constraint on the number for each character of the string). The non-constructive one has some overflow problems with radixes which are powers of 2. It's also slow for radixes which aren't powers of 2.

owen-jones-diffblue commented:
@romainbrenguier @allredj Please review

owen-jones-diffblue commented:
This PR is for issue https://github.com/diffblue/test-gen/issues/863

allredj commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -234,35 +238,34 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
-      get_numeric_value_from_character(res[0], char_type, type, radix_ul);
-
-    for(size_t j=1; j<size; j++)
+    if(size>1)
```
The documentation above (`sum_j`, `a5`) must be updated if we no longer iterate over `j`. At the moment I'm having difficulties evaluating the soundness of this new method. Do you mind updating the doc before I come back to reviewing this?

owen-jones-diffblue replied:
I've done that. Apologies for leaving out-of-date comments. The main thing to notice is that `sum` and `digit_constraints` persist between iterations of the loop, so we only need to add a little to them to make them correct for the new value of `size`. And you can think of `j` as `size-1` - it's the index of the last character if `|res| == size`.

romainbrenguier commented on regression/strings-smoke-tests/java_long_to_string/Test1.java:
```diff
@@ -0,0 +1,9 @@
+    {
+        String s = Long.toString(12);
+        assert(s.equals("12"));
+        assert(!s.equals("12"));
```
Having both assertions in the same branch can be a problem because CBMC will assume the result of the first one is true when we get to the second one, so it would be trivially false in that case.
You can add an `if` with a boolean parameter that can be an argument of the function.

romainbrenguier commented on regression/strings-smoke-tests/java_long_to_string/Test2.java:
```diff
@@ -0,0 +1,9 @@
+    {
+        String s = Long.toString(-23);
+        assert(s.equals("-23"));
+        assert(!s.equals("-23"));
```
same remark

romainbrenguier commented on regression/strings-smoke-tests/java_long_to_string/Test5.java:
```diff
@@ -0,0 +1,9 @@
+    {
+        String s = Long.toString(0);
+        assert(s.equals("0"));
+        assert(!s.equals("0"));
```
same remark

romainbrenguier commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -320,14 +320,24 @@ class string_constraint_generatort: messaget
     const std::size_t max_string_length,
-    const exprt& radix);
+    const exprt& radix,
+    const unsigned long radix_ul=0);
```
This is a lot of arguments. You could recover `char_type` and `index_type` from the type of `str` so they are redundant.

owen-jones-diffblue replied:
Done

romainbrenguier commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -320,14 +320,24 @@ class string_constraint_generatort: messaget
+    const string_exprt& str,
+    const std::size_t max_string_length,
+    const exprt& radix,
+    const unsigned long radix_ul=0);
```
same remark

romainbrenguier commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -365,15 +375,16 @@ class string_constraint_generatort: messaget
 
-exprt is_digit_with_radix(const exprt &chr, const exprt &radix);
+exprt is_digit_with_radix(
+  const exprt &chr, const exprt &radix_char_type, unsigned long radix=0);
```
I find it confusing that `radix_char_type` is not a `typet`. Maybe change it to `radix_as_char`.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -151,11 +151,13 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
+  /// Most of the time we can evaluate radix as an integer. The value 0 is used
+  /// to indicate when we can't tell what the radix is.
+  const unsigned long radix_ul=to_integer_or_default(radix, 0);
+  PRECONDITION((radix_ul>=2 && radix_ul<=36) || radix_ul==0);
```
It does not seem like a precondition to me, I think `CHECK_RETURN` would be more appropriate.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -151,11 +151,13 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
-    max_size=max_printed_string_length(type, to_integer_or_default(radix, 2ul));
+    max_size=max_printed_string_length(type, radix_ul);
   }
   PRECONDITION(max_size<std::numeric_limits<size_t>::max());
```
I prefer to have preconditions at the beginning of the function.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -207,64 +202,41 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
+      res[j],
+      zero_char);
+    axioms.push_back(
+      is_digit_with_radix_lower_case(chr, radix_char_type, radix_ul));
```
can we pass the call inside the `if`? What I mean is `j<res.length?is_digit(chr):true` instead of `is_digit(j<res.length?chr:'0')`. It seems to me the generated formula would be slightly simpler.

owen-jones-diffblue replied:
I think the reason I'm not doing this is that you end up having `j<=10?is_digit(str[10-j]):true` and it doesn't like having 'str[10-j]` when `j>10`, even though it will never be evaluated in that case. I'll try it again and double check that this is true in this case.

owen-jones-diffblue replied:
Oh, it does seem to work. I was being overly cautious. I know there are some other situations where it really can't cope with that kind of thing.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -207,64 +202,41 @@ string_exprt string_constraint_generatort::add_axioms_from_int_with_radix(
-      and_exprt(premise, starts_with_minus),
-      equal_exprt(input_int, unary_minus_exprt(sum)));
-    axioms.push_back(a7);
+//  if(radix_ul==2ul || radix_ul==4ul || radix_ul==8ul || radix_ul==16ul)
```
use `#if 0` for commenting code and explain why it is commented.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -512,49 +587,62 @@ exprt string_constraint_generatort::add_axioms_for_characters_in_integer_string(
-        no_overflow);
+      if(radix_ul==0)
+      {
+        radix_to_power_of_i=power_exprt(radix, i_expr);
```
doesn't power_exprt work only for `float`? You can use `mult_exprt(radix_to_power_of_i, radix)` instead

owen-jones-diffblue replied:
It works fine, and it's faster than what you suggest. If you really want I can remove it.

romainbrenguier replied:
I tried to write some with integer variables but the solver would give incorrect results (looking at bits in the result, it looked like a float).
Maybe it works here because `radix` and `i_expr` are constants but it's better if we can avoid relying on that.

owen-jones-diffblue replied:
We don't use `power_exprt` any more, because the constructive approach turned out to be better than the non-constructive approach.

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -512,49 +587,62 @@ exprt string_constraint_generatort::add_axioms_for_characters_in_integer_string(
+      else
+      {
+        radix_to_power_of_i=from_integer(
+          static_cast<unsigned long>(pow(radix_ul, i)), type);
```
same remark

owen-jones-diffblue replied:
Since C++11 there have been versions of `pow()` for all arithmetic types (they cast the arguments to double internally). Are you worried about floating point errors?

owen-jones-diffblue replied:
I've changed this

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -571,92 +659,107 @@ exprt string_constraint_generatort::add_axioms_for_parse_int(
-/// \param radix:  the radix
+/// \param radix_char_type:  the radix as an expression of the same type as chr
+/// \param radix: the radix, which should be between 2 and 36 (default 0, which
+///   is treated specially)
```
specially how?

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -571,92 +659,107 @@ exprt string_constraint_generatort::add_axioms_for_parse_int(
 /// \param radix_char_type:  the radix as an expression of the same type as chr
-/// \param radix: the radix, which should be between 2 and 36 (default 36)
+/// \param radix: the radix, which should be between 2 and 36 (default 0, which
+///   is treated specially)
```
same remark

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -734,12 +838,16 @@ exprt get_numeric_value_from_character(
 /// return a value 1 larger than needed, which is fine for our purposes.
 /// \param type: the type that we are considering values of
-/// \param ul_radix: the radix we are using, as a double
+/// \param radix_ul: the radix we are using [default 0]
```
what does it return for 0?

owen-jones-diffblue commented:
@romainbrenguier I've addressed all of your comments except for the one about `power_exprt`

allredj commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -466,42 +439,145 @@ void string_constraint_generatort::add_axioms_for_correct_number_format(
-  const symbol_exprt& x,
+/// \param radix_ul: the radix as an unsigned long, or 0 if that can't be
+///   determined
+void string_constraint_generatort::add_axioms_for_characters_in_integer_string_constructive(
```
The documentation should be adapted to the new list of parameters.

owen-jones-diffblue replied:
The constructive approach won, which means the function name now fits on one line. It also has the side effect that we don't use `power_exprt` any more.

allredj commented on unit/solvers/refinement/string_constraint_generator_valueof/is_digit_with_radix_lower_case.cpp:
```diff
@@ -15,43 +15,99 @@
 
-SCENARIO("is_digit_with_radix_lower_case",
-  "[core][solvers][refinement][string_constraint_generator_valueof]")
+static exprt actual(const mp_integer int_value, const unsigned long radix_ul)
```
Could we have a one line description of what that function does? Same for `actual_with_radix`.

owen-jones-diffblue commented:
@allredj I've addressed all of your comments except the extremely long function name and `power_exprt`, both of which I'm looking into

owen-jones-diffblue commented:
I've addressed all comment except having a test that uses `Integer` (and also `Long`). I'm going to improve the tests a bit now, but I don't plan to change the code so you can start reviewing it again @romainbrenguier and @allredj.

owen-jones-diffblue commented:
@thk123  If you could have a look at this that would be great.

allredj commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -419,142 +334,182 @@ string_exprt string_constraint_generatort::add_axioms_for_value_of(
-  exprt radix_to_power_of_i;
-  exprt no_overflow;
+  exprt sum=get_numeric_value_from_character(
+      str[0], char_type, type, strict_formatting, radix_ul);
```
bad indent

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -363,18 +370,21 @@ class string_constraint_generatort: messaget
     exprt char1, exprt char2, exprt char_a, exprt char_A, exprt char_Z);
   bool is_constant_string(const string_exprt &expr) const;
   string_exprt empty_string(const refined_string_typet &ref_type);
+  unsigned long to_integer_or_default(const exprt &expr, unsigned long def);
```
Why do we use `unsigned long` here?

owen-jones-diffblue replied:
Because `mp_integer` has a `integer2ulong` method. Valid radixes should be between 2 and 36, but I wrote this function to be more general. I'm happy to change the return value if you like.

allredj replied:
It's fine I guess. I think in other places we stuck to `mp_integer` for consistency, but whatever you feel is better.

owen-jones-diffblue commented:
I've addressed Joel's comments.

I'll probably squash everything into one commit before it's merged, because I'm not sure the commit history is very helpful in this case.

owen-jones-diffblue commented:
@peterschrammel Please review

owen-jones-diffblue commented:
This is ready except for having a senior reviewer. @mgudemann? @peterschrammel? @thk123?

romainbrenguier commented on src/solvers/refinement/string_constraint_generator_valueof.cpp:
```diff
@@ -529,107 +486,120 @@ exprt string_constraint_generatort::add_axioms_for_parse_int(
+/// \param expr: input expression
+/// \param def: default value to return if we cannot evaluate expr
+/// \return the output as an unsigned long
+unsigned long string_constraint_generatort::to_integer_or_default(
```
this can be made static if we pass `ns` as an argument
